src/cdk/a11y/focus-trap/configurable-focus-trap-factory.ts

```diff
@@ -0,0 +1,52 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {DOCUMENT} from '@angular/common';
+import {
+  Inject,
+  Injectable,
+  Optional,
+  NgZone,
+} from '@angular/core';
+import {InteractivityChecker} from '../interactivity-checker/interactivity-checker';
+import {ConfigurableFocusTrap} from './configurable-focus-trap';
+import {FOCUS_TRAP_INERT_STRATEGY, FocusTrapInertStrategy} from './focus-trap-inert-strategy';
+import {EventListenerFocusTrapInertStrategy} from './event-listener-inert-strategy';
+import {FocusTrapManager} from './focus-trap-manager';
+
+/** Factory that allows easy instantiation of configurable focus traps. */
+@Injectable({providedIn: 'root'})
+export class ConfigurableFocusTrapFactory {
+  private _document: Document;
+  private _inertStrategy: FocusTrapInertStrategy;
+
+  constructor(
+      private _checker: InteractivityChecker,
+      private _ngZone: NgZone,
+      private _focusTrapManager: FocusTrapManager,
+      @Inject(DOCUMENT) _document: any,
+      @Optional() @Inject(FOCUS_TRAP_INERT_STRATEGY) _inertStrategy?: FocusTrapInertStrategy) {
+
+    this._document = _document;
+    this._inertStrategy = (_inertStrategy ? _inertStrategy :
+      new EventListenerFocusTrapInertStrategy());
```

Comment:
Having people pass in the strategy instance directly is probably our best bet.

Reply:
I was hoping to migrate cdkTrapFocus to ConfigurableFocusTrap under the hood. In order to avoid always including the EventListenerFocusTrapInertStrategy code, cdkTrapFocus would have to require a strategy instance too. I suppose it could default to the old FocusTrap if no strategy was provided, but ideally I had wanted it to default to EventListenerFocusTrapInertStrategy. It sounds like maybe a default strategy isn't feasible, and developers won't be migrated to the new behavior until they explicitly opt-in to a strategy?

Reply:
stevenyxu@ brought up that maybe we could have ConfigurableFocusTrapFactory, which requires a strategy, and ConfigurableFocusTrapFactoryWithDefaults, which brings in EventListenerFocusTrapInertStrategy. We could migrate cdkTrapFocus and any components that we want to force-migrate to ConfigurableFocusTrapFactoryWithDefaults, and then anyone who wants a custom strategy could use ConfigurableFocusTrapFactory.

Reply:
I think we had a similar problem for the `DateAdapter`. In that case we wound up creating different modules people could import for the different strategies: `MatNativeDateModule`, `MatMomentDateModule`, etc. and threw an error in the datepicker telling them to include one of those modules if we detected that the `DateAdapter` was null. This wouldn't allow the under the hood migration though, because it still requires users to choose a module to import

Reply:
@jelbourn what do you think? EventListenerFocusTrapInertStrategy is not much code, but I do want to be consistent with the rest of the library. What would you like to see happen with the cdkTrapFocus migration? It's always an option to make a new directive...

Reply:
Yeah, it is a pretty small amount of code. I'm good with leaving it as is and adding a TODO to split up the strategies into different modules in a later Angular version.

Reply:
Sounds good! I added a TODO.